### PR TITLE
GAWB 1505: expose api that returns the list of groups a user belongs to

### PR DESCRIPTION
**Pull request by ahaessly**
- [x] **Submitter**: Include the JIRA issue number in the PR description
- [ ] **Submitter**: Check that the **Product Owner** has signed off on any user-facing changes
- [x] **Submitter**: Make sure Swagger is updated if API changes
  - [ ] **...and Orchestration's Swagger too!**
- [ ] **Submitter**: If updating admin endpoints, also update [firecloud-admin-cli](https://github.com/broadinstitute/firecloud-admin-cli)
- [x] **Submitter**: Check documentation and code comments. Add explanatory PR comments if helpful.
- [x] **Submitter**: JIRA ticket checks:
  * Acceptance criteria exists and is met
  * Note any changes to implementation from the description
  * Add notes on what you've tested
- [x] **Submitter**: Update RC_XXX release ticket with any config or environment changes necessary
- [x] **Submitter**: Database checks:
  * If PR includes new or changed db queries, include the explain plans in the description
  * Make sure liquibase is updated if appropriate
  * If doing a migration, take a backup of the
  [dev](https://console.cloud.google.com/sql/instances/terraform-qfarbdq3lrexxck5htofjs5z6m/backups?project=broad-dsde-dev&organizationId=548622027621)
  and
  [alpha](https://console.cloud.google.com/sql/instances/terraform-r4caezzc35c4tb7pgdhwkmme4y/backups?project=broad-dsde-alpha&organizationId=548622027621)
  DBs in Google Cloud Console
- [x] **Submitter**: Update FISMA documentation if changes to:
  * Authentication
  * Authorization
  * Encryption
  * Audit trails
- [x] **Submitter**: If you're adding new libraries, sign us up to security updates for them
- [ ] Tell your tech lead (TL) that the PR exists if they want to look at it
- [x] Anoint a lead reviewer (LR). **Assign PR to LR**
* Review cycle:
  * LR reviews
  * Rest of team may comment on PR at will
  * **LR assigns to submitter** for feedback fixes
  * Submitter rebases to develop again if necessary
  * Submitter makes further commits. DO NOT SQUASH
  * Submitter updates documentation as needed
  * Submitter **reassigns to LR** for further feedback
- [ ] **TL** sign off
- [x] **LR** sign off
- [ ] **Assign to submitter** to finalize
- [ ] **Submitter**: Verify all tests go green, including CI tests
- [ ] **Submitter**: Squash commits and merge to develop
- [ ] **Submitter**: Delete branch after merge
- [ ] **Submitter**: **Test this change works on dev environment after deployment**. YOU own getting it fixed if dev isn't working for ANY reason!
- [ ] **Submitter**: Verify swagger UI on dev environment still works after deployment
- [ ] **Submitter**: Inform other teams of any API changes via hipchat and/or email
- [ ] **Submitter**: Mark JIRA issue as resolved once this checklist is completed

### CODE REVIEW

File: core/src/test/scala/org/broadinstitute/dsde/rawls/dataaccess/slick/RawlsUserComponentSpec.scala [hunk 1]
```diff
@@ -80,4 +80,36 @@ class RawlsUserComponentSpec extends TestDriverComponentWithFlatSpecAndMatchers
...
+
+    runAndWait(rawlsUserQuery.save(user1))
+    runAndWait(rawlsUserQuery.save(user2))
+    runAndWait(rawlsGroupQuery.save(group1))
+    runAndWait(rawlsGroupQuery.save(group2))
+    runAndWait(rawlsGroupQuery.save(group3))
+    runAndWait(rawlsGroupQuery.save(group4))
+    runAndWait(rawlsGroupQuery.save(group5))
+
+    assertResult(Set(RawlsGroupRef(RawlsGroupName("Group Two")),RawlsGroupRef(RawlsGroupName("Group Two")),RawlsGroupRef(RawlsGroupName("Group Three")),RawlsGroupRef(RawlsGroupName("Group Five")))) {
```
Comment (helgridly) [inline core/src/test/scala/org/broadinstitute/dsde/rawls/dataaccess/slick/RawlsUserComponentSpec.scala]: Am I misreading something here? This looks like a set with Group Two twice, which is not what a set is.

File: core/src/test/scala/org/broadinstitute/dsde/rawls/dataaccess/slick/RawlsUserComponentSpec.scala [hunk 2]
```diff
@@ -80,4 +80,36 @@ class RawlsUserComponentSpec extends TestDriverComponentWithFlatSpecAndMatchers
...
+    val email2 = RawlsUserEmail("two.emails@are.better.than.one")
+    val user2 = RawlsUser(subjId2, email2)
+
+
+    //    val testUser2 = testUser.copy(userSubjectId = RawlsUserSubjectId("dummy-ID2"), userEmail = RawlsUserEmail("dummy-email2@example.com"))
+    val group1 = makeRawlsGroup("Group One", Set(user1))
+    val group2 = makeRawlsGroup("Group Two", Set(user2)).copy(subGroups = Set(group1))
+    val group3 = makeRawlsGroup("Group Three", Set.empty).copy(subGroups = Set(group2))
+    val group4 = makeRawlsGroup("Group Four", Set.empty).copy(subGroups = Set(group1))
+    val group5 = makeRawlsGroup("Group Five", Set.empty).copy(subGroups = Set(group3))
```
Comment (helgridly) [inline core/src/test/scala/org/broadinstitute/dsde/rawls/dataaccess/slick/RawlsUserComponentSpec.scala:109]: maybe put the graph in a comment here, so my brain doesn't explode

```
G5 - G3 - G2 - G1 - U1
            \- U2

G4 - G1 - U1
```

File: core/src/test/scala/org/broadinstitute/dsde/rawls/dataaccess/slick/RawlsUserComponentSpec.scala [hunk 3]
```diff
@@ -80,4 +80,36 @@ class RawlsUserComponentSpec extends TestDriverComponentWithFlatSpecAndMatchers
...
+    runAndWait(rawlsGroupQuery.save(group1))
+    runAndWait(rawlsGroupQuery.save(group2))
+    runAndWait(rawlsGroupQuery.save(group3))
+    runAndWait(rawlsGroupQuery.save(group4))
+    runAndWait(rawlsGroupQuery.save(group5))
+
+    assertResult(Set(RawlsGroupRef(RawlsGroupName("Group Two")),RawlsGroupRef(RawlsGroupName("Group Two")),RawlsGroupRef(RawlsGroupName("Group Three")),RawlsGroupRef(RawlsGroupName("Group Five")))) {
+      runAndWait(rawlsGroupQuery.listGroupsForUser(user2))
+    }
+  }
```
Comment (helgridly) [inline core/src/test/scala/org/broadinstitute/dsde/rawls/dataaccess/slick/RawlsUserComponentSpec.scala:128]: could do with a test for when the answer is "no groups", too

File: core/src/test/scala/org/broadinstitute/dsde/rawls/dataaccess/slick/RawlsUserComponentSpec.scala (hunk 1 above)
Comment (helgridly) [reply core/src/test/scala/org/broadinstitute/dsde/rawls/dataaccess/slick/RawlsUserComponentSpec.scala]: Oh, I see. I thought it was a typo for Group One, but it's not - it's just duplicated. And then the call to set collapses it anyway. Can you chuck one of them?

Comment (ahaessly) [reply core/src/test/scala/org/broadinstitute/dsde/rawls/dataaccess/slick/RawlsUserComponentSpec.scala]: oops

File: core/src/test/scala/org/broadinstitute/dsde/rawls/dataaccess/slick/RawlsUserComponentSpec.scala (hunk 3 above)
Comment (ahaessly) [reply core/src/test/scala/org/broadinstitute/dsde/rawls/dataaccess/slick/RawlsUserComponentSpec.scala:128]: will do

File: core/src/main/resources/swagger/rawls.yaml [hunk 4]
```diff
@@ -3308,6 +3308,21 @@ paths:
             - openid
             - email
             - profile
+  '/api/user/groups':
+    get:
+      responses:
+        '200':
+          description: Successful Request
+      description: "get the list of groups the user belong to"
```
Comment (davidangb) [inline core/src/main/resources/swagger/rawls.yaml]: nit: don't need quotes; also, "belong" -> "belongs"

for completeness, specify the response structure, something like:
```
        '200':
          description: Successful Request
          schema:
            type: array
            description: list of the user's groups
            items:
              type: string
```

**Review by davidangb — COMMENTED** [review]
minor swagger comments inline.

**Comment by helgridly** [comment]
👍

File: core/src/main/scala/org/broadinstitute/dsde/rawls/webservice/UserApiService.scala [hunk 5]
```diff
@@ -90,6 +90,13 @@ trait UserApiService extends HttpService with PerRequestCreator with UserInfoDir
           UserService.ListBillingAccounts)
       }
     } ~
+      path("user" / "groups") {
```
Comment (dvoet) [inline core/src/main/scala/org/broadinstitute/dsde/rawls/webservice/UserApiService.scala:93]: write a test for this api

**Review by davidangb — APPROVED** [review]
👍 (modulo Doug's comment)

**Comment by helgridly** [comment]
👍 (again)

**Pull request merged**